### DISCUSSION TITLE
File based delta update

### DISCUSSION THREAD

**Discussion by minecraft2048** [rauc/rauc#1065] [Q&A]
Lets say on my update server I have `prev_version.raucb` and `current_version.raucb`. Is there some sort of way to generate binary diff between prev version and current version, pack it into a patch file, send it to the hardware, and apply the patch?

Something like `rauc diff prev_version.raucb current_version.rauc diff.raucb_delta` then to install it you run `rauc install diff.rauc_delta`

The closest thing that RAUC have is adaptive update and casync, and both of them assumes that the hardware and update server can communicate through HTTP. The problem is that I can only transfer file to the hardware

**Comment by ejoerns** [comment]
@minecraft2048 There is no real binary or file-based diff support in RAUC, yet. It could be added as another adaptive update method but that would still rely on HTTP transport.

How do you bring the files to your target? I assume that you use some sort of resource-limited communication channel if asking for delta updates?

**Reply by minecraft2048** [reply]
Some context: The device will be a payload on a cubesat, and the provider told me that they can probably give me 500 kB/s uplink

The updates will be transferred over some sort of file transfer protocol to another embedded linux computer, then I can transfer it to my device using standard tools such as scp or wget

**Reply by jluebbe** [reply]
Very interesting!

Your suggested command is very close to what we have on our list of things to implement when we have time or a customer who needs it. The delta bundle would contain the checksums for the "base version", so we could check that an update is only applied to a matching base. For generating deltas, we could generate and store rsync "batch files" in the delta bundle, which work well with tar archives or properly tuned read-only images.
